### PR TITLE
fix dwb controller

### PR DESCRIPTION
**Pull request by daisukes**
- extends dwb_local_planner and override the `transformGlobalPlan` method
- the default method checks poses in the plan by the Euclidean distance from the robot pose to determine the local goal target
- new method checks accumulated distance from the robot pose

### CODE REVIEW

**Comment by daisukes** [comment]
@muratams

Could you check the code?

**Comment by muratams** [comment]
diagram

File: cabot_navigation2/plugins/cabot_dwb_local_planner.cpp [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      return euclidean_distance(pose1, robot_pose.pose) < euclidean_distance(pose2, robot_pose.pose);
+    });
+  double nearest_distance = euclidean_distance(*nearest_point_from_robot, robot_pose.pose);
+
+  // Customized: Find the first pose less than double of nearset_distance - adhoc solution
+  auto search_start_point = std::find_if(
+    global_plan_.poses.begin(), global_plan_.poses.end(),
+    [&](const auto & pose) {
+      return euclidean_distance(pose, robot_pose.pose) < nearest_distance * 2;
+    });
```
Comment (muratams) [inline cabot_navigation2/plugins/cabot_dwb_local_planner.cpp:135]: What is the purpose of adjusting search_start_point from nearest_point_from_robot?

**Review by muratams — APPROVED** [review]
Confirmed that this pull request solved the issue described in the first comment.
Can be merged to main.

File: cabot_navigation2/plugins/cabot_dwb_local_planner.cpp (hunk 1 above)
Comment (daisukes) [reply cabot_navigation2/plugins/cabot_dwb_local_planner.cpp:135]: In the original logic, the path's poses are observed from the beginning, and the process stops when the pose is farther than the threshold distance. There was an issue where if the initial pose is within the threshold circle, but the path subsequently moves outside the circle, the target point was set to the first pose outside the circle. (If the red line in your figure starting within the circle.)

Screenshot 2024-07-03 at 11 54 25 PM

While drawing this, I found that the current logic may not work in some cases...

**Comment by daisukes** [comment]
image

How about new logic to find the start?

**Comment by muratams** [comment]
The new implementation is better than the previous (original) implementation in many cases.

**Comment by muratams** [comment]
https://github.com/CMU-cabot/cabot-navigation/pull/41#issuecomment-2207965878

The distance between the search_start_point and the robot_pose (2D in the figure) was replaced with `transform_start_threshold / 2` by the two follow-up commits.

**Pull request merged**